File: app/src/main/java/com/example/musicartistsapp/FilterFragment.java
package com.example.musicartistsapp;

import android.content.Context;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import androidx.fragment.app.DialogFragment;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.example.musicartistsapp.databinding.FilterFragmentBinding;

public class FilterFragment extends DialogFragment implements View.OnClickListener {

    interface FilterListener {

        void onFilter(FilterModel filter);
    }
    private FilterFragmentBinding filterFragmentBinding;

    private FilterListener filterListener;
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        filterFragmentBinding = FilterFragmentBinding.inflate(inflater, container, false);

        filterFragmentBinding.buttonSearch.setOnClickListener(this);
        filterFragmentBinding.buttonCancel.setOnClickListener(this);

        return filterFragmentBinding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        filterFragmentBinding = null;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof FilterListener) {
            filterListener = (FilterListener) context;
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        //why this code exists?
        getDialog().getWindow().setLayout(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
    }

    public void onSearchClicked() {
        if (filterListener != null) {
            filterListener.onFilter(createFilter());
        }
        dismiss();
    }

    private String retrieveSelectedCountry() {
        String selectedCountry = (String) filterFragmentBinding.spinnerCountry.getSelectedItem();
        return getString(R.string.any).equals(selectedCountry) ? null : selectedCountry;
    }

    private String retrieveSelectedGenre() {
        String selectedGenre = (String) filterFragmentBinding.sprinnerGenre.getSelectedItem();
        return getString(R.string.any).equals(selectedGenre) ? null : selectedGenre;
    }

    private FilterModel createFilter() {
        FilterModel filter = new FilterModel();
        filter.setCountry(retrieveSelectedCountry());
        filter.setGenre(retrieveSelectedGenre());
        return filter;
    }

    public void onCancelClicked() {
        dismiss();
    }

    public void setDefaultSelection() {
        if (filterFragmentBinding != null) {
            filterFragmentBinding.spinnerCountry.setSelection(0);
            filterFragmentBinding.sprinnerGenre.setSelection(0);
        }
    }


    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.button_search:
                onSearchClicked();
                break;
            case R.id.button_cancel:
                onCancelClicked();
                break;
        }
    }
}

File: app/src/main/java/com/example/musicartistsapp/FilterFragment.kt
package com.example.musicartistsapp

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import com.example.musicartistsapp.databinding.FilterFragmentBinding
import java.lang.IllegalStateException

class FilterFragment : DialogFragment(), View.OnClickListener {
    interface FilterListener {
        fun onFilter(filter: FilterModel)
    }

    private lateinit var filterFragmentBinding: FilterFragmentBinding
    private lateinit var filterListener: FilterListener

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        filterFragmentBinding = FilterFragmentBinding.inflate(inflater, container, false)

        filterFragmentBinding.buttonSearch.setOnClickListener(this)
        filterFragmentBinding.buttonCancel.setOnClickListener(this)
        return filterFragmentBinding.root
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is FilterListener) {
            filterListener = context
        }
    }

    override fun onResume() {
        super.onResume()
        val window = dialog?.window ?: throw IllegalStateException("Cannot retrieve dialog window")
        window.setLayout(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT)
    }

    private fun onSearchClicked() {
        filterListener.onFilter(createFilter())
        dismiss()
    }

    private fun retrieveSelectedCountry(): String? {
        val selectedCountry = filterFragmentBinding.spinnerCountry.selectedItem as String
        return if (getString(R.string.any) == selectedCountry) null else selectedCountry
    }

    private fun retrieveSelectedGenre(): String? {
        val selectedGenre = filterFragmentBinding.sprinnerGenre.selectedItem as String
        return if (getString(R.string.any) == selectedGenre) null else selectedGenre
    }

    private fun createFilter(): FilterModel {
        val filter = FilterModel()
        filter.country = retrieveSelectedCountry()
        filter.genre = retrieveSelectedGenre()
        return filter
    }

    private fun onCancelClicked() {
        dismiss()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.button_search -> onSearchClicked()
            R.id.button_cancel -> onCancelClicked()
        }
    }
}